### ISSUE TITLE
Recipe ReplaceLambdaWithMethodReference results in non compilable code cause it does not added an import for a needed class. 

### ISSUE THREAD

**Issue by jbessels** [openrewrite/rewrite#3071]
Running ReplaceLambdaWithMethodReference in this simpliefied example results in code which no longer compiles

**Original code**
```
String fullName = Optional.ofNullable("Turing")
		.map(n -> UserManager.INSTANCE.getUser(n))
		.map(user -> user.getFullName())
		.orElse("");
```
	
**Converted code**					
```
String fullName = Optional.ofNullable("Turing")
		.map(UserManager.INSTANCE::getUser)
		.map(User::getFullName)
		.orElse("");
```
		
It does not compile because a 'import User' line is missing. In the original code UserManager.INSTANCE.getUser(n) returns an object of type User. No import for it is needed. In the converted code now the class User is explicit used and a method in it called. An 'import class' in IntelliJ fixes the problem. Hence ReplaceLambdaWithMethodReference should create an import when needed to fix this.

**Comment by knutwannheden** [comment]
@jbessels Thanks for the report! I was able to reproduce the error using this test class:
```java
              import java.net.URI;
              import java.nio.file.Paths;
              import java.util.Optional;

              class A {
                  void m() {
                      URI uri = Optional.ofNullable("path")
                            .map(Paths::get)
                            .map(path -> path.toUri())
                            .orElse(null);
                  }
              }
```

**Comment by jbessels** [comment]
Thanks, that it is fixed. The majority (85%+) of 'breaking code aka not compiling' code after running recipes has to do with Lambda's for our organic/exotic repo.Another repo had very little Lambda issues.  The nasty repo  uses Wicket (web framework, where often 6 lines of code can be replaced by 1 Lambda line)  Ambiguity issues (more tickets incoming) is a real can of worms. And as you said in another ticket 'surprisingly difficult to detect'. Atm I have to exclude about 40 files wrt Lambda edge case.  Some issues affect 8+ files.

When will this fix finds its way in the 4.43.0-SNASHOT - which uses 7.39.0-SNAPSHOT?? **Aka when can I test it?**

**Comment by jbessels** [comment]
Not sure if comments in closed tickets are read. Just reran OpenRewrite again and it **still fails. No import for User is added.**

        
        <rewrite-maven-plugin.version>4.45.0</rewrite-maven-plugin.version>
        <rewrite-migrate-java.version>1.21.1</rewrite-migrate-java.version>
        <rewrite-testing-frameworks.version>1.37.0</rewrite-testing-frameworks.version>

**Comment by knutwannheden** [comment]
@jbessels That is odd. Is this in a public repo somewhere? It looks similar enough to the test case I added, yet there must be a subtle difference somewhere...

**Comment by jbessels** [comment]
Yes must be. I've double checked and it still fails. Can you point me to your test case and the changes made to code?

**Comment by knutwannheden** [comment]
@jbessels Here is the test that I added: https://github.com/openrewrite/rewrite/blob/b9a0475eb72a5cdde5003a64f57bae0a9bf08cb4/rewrite-java-test/src/test/java/org/openrewrite/java/cleanup/ReplaceLambdaWithMethodReferenceTest.java#L966-L1003

**Comment by jbessels** [comment]
@knutwannheden Missed your response. It DOES work with your example. **Not with my code.** 
 
I've inserted your test code in the same class as where the User import was not added. Your Path import was correctly added. Paths and Path are in the same Java package. This got me thinking. class User is from another module (with another package name). Because running openRewrite on the whole big 100++ Maven modules repo is painfully slow I ran it for the module where it did go wrong. Hence the complete LST for the repo is NOT available and class User is unknown. **I hoped that when running it for the whole repo it would work. Unfortunately NOT.** 

I checked and there is only one User.java class just to be sure that it would not break with multiple classes of the same name. Beneath a screenshot of  what IntelliJ shows. Java knows that the line above user -> user.getFullName() results in an User object. 

![image](https://user-images.githubusercontent.com/69236284/236745554-93d8ad24-08b2-42cd-a8ff-5d47078d149c.png)

**Comment by knutwannheden** [comment]
@jbessels Thanks a lot for your investigation! It would seem like the recipe needs to make sure that the type attribution is complete before making the substitution. I will report back when I know more.

**Comment by knutwannheden** [comment]
@jbessels I took another look at the code and still can't quite figure it out. What I did find however, is that the recipe doesn't properly deal with nested classes. So if your `User` class for example is a nested class in `UserManager`, I can see how this could go wrong. Can you check whether that is the case?

**Comment by jbessels** [comment]
Did some digging. UserManager is an enum. No nested class in it.
`public enum UserManager implements Caching { .... }`

User is a normal class
`public class User extends CRUDObject<User> implements Comparable<User>, Serializable {`

If needed I can send you the involved  classes in a DM Slack.

**Comment by knutwannheden** [comment]
@jbessels Yes, it would be great if you could send me all I need to reproduce the problem.

**Comment by jbessels** [comment]
Will send it all this evening. That pesky sprint work is in the way ;-)

**Comment by jbessels** [comment]
I've retested it. Somewhere between last comment in the thread and the latest version of the plugin it got solved. As in the User object is added. See screenshot. **Hence 3071 can be closed.**

![image](https://github.com/openrewrite/rewrite/assets/69236284/c9eb1ed5-2c6e-44e3-aee2-b84d5f16ec88)
